closed/make/DDR.gmk

```diff
@@ -0,0 +1,122 @@
+# 2 along with this work; if not, see <http://www.gnu.org/licenses/>.
+# ===========================================================================
+
+default :
+	$(error DDR.gmk has no default target)
+
+ifeq (,$(wildcard $(SPEC)))
+  $(error DDR.gmk needs SPEC set to a proper spec.gmk)
+endif
+
+include $(SPEC)
+include $(TOPDIR)/make/common/MakeBase.gmk
+include $(TOPDIR)/make/common/JavaCompilation.gmk
+include $(TOPDIR)/make/common/SetupJavaCompilers.gmk
+
+# Supporting definitions.
+DDR_SUPPORT_DIR := $(SUPPORT_OUTPUTDIR)/ddr
+DDR_VM_SRC_ROOT := $(OPENJ9_TOPDIR)/debugtools/DDR_VM/src
+
+#############################################################################
+
+# Build the class generator.
+$(eval $(call SetupJavaCompilation,BUILD_DDR_TOOLS, \
+	SETUP := GENERATE_OLDBYTECODE, \
+	BIN := $(DDR_SUPPORT_DIR)/tools, \
+	CLASSPATH := $(addprefix $(JDK_OUTPUTDIR)/modules/, java.base openj9.dtfj), \
+	SRC := $(DDR_VM_SRC_ROOT), \
+	INCLUDE_FILES := $(addsuffix .java,$(subst .,/, \
+		com.ibm.j9ddr.BytecodeGenerator \
+		com.ibm.j9ddr.CTypeParser \
+		com.ibm.j9ddr.StructureHeader \
+		com.ibm.j9ddr.StructureReader \
+		com.ibm.j9ddr.StructureTypeManager \
+		com.ibm.j9ddr.logging.LoggerNames \
+		com.ibm.j9ddr.tools.ClassGenerator \
+		com.ibm.j9ddr.tools.FlagStructureList \
+		com.ibm.j9ddr.tools.store.J9DDRStructureStore \
+		com.ibm.j9ddr.tools.store.StructureKey \
+		com.ibm.j9ddr.tools.store.StructureMismatchError \
+		)), \
+	))
+
+#############################################################################
+
+# The blob is built with the VM.
+DDR_BLOB_FILE := $(OUTPUTDIR)/vm/j9ddr.dat
+
+# Generate pointer and structure classes.
+DDR_GENERATED_MARKER := $(DDR_SUPPORT_DIR)/generated.done
+
+# $(DDR_VM_SRC_ROOT) is on the classpath as it contributes StructureAliases29.dat
+DDR_GENERATOR_CLASSPATH := $(call PathList, $(BUILD_DDR_TOOLS_BIN) $(DDR_VM_SRC_ROOT))
+
+$(DDR_GENERATED_MARKER) : $(DDR_BLOB_FILE) $(BUILD_DDR_TOOLS)
+	@$(ECHO) Generating DDR pointer and structure classes
+	@$(RM) -rf $(DDR_SUPPORT_DIR)/generated
+	@$(JAVA) -cp $(DDR_GENERATOR_CLASSPATH) \
+		com.ibm.j9ddr.tools.ClassGenerator \
+			--blob=$(DDR_BLOB_FILE) \
+			--out=$(DDR_SUPPORT_DIR)/generated
+	@$(TOUCH) $@
+
+generate : $(DDR_GENERATED_MARKER)
+
+#############################################################################
+
+# Supporting definitions.
+DDR_CLASSES_BIN := $(DDR_SUPPORT_DIR)/classes
+
+# We depend upon the generated classes.
+DDR_CLASSPATH := $(DDR_SUPPORT_DIR)/generated
+
+# We also depend upon class files from these modules.
+DDR_CLASSPATH += $(addprefix $(JDK_OUTPUTDIR)/modules/, \
+		java.base \
+		java.desktop \
+		openj9.dtfj \
+		openj9.traceformat \
+	)
+
+DDR_SRC_EXCLUDES := com/ibm/j9ddr/tools/ant
+
+ifeq (,$(filter mz31 mz64,$(OPENJ9_PLATFORM_CODE)))
+DDR_SRC_EXCLUDES += com/ibm/j9ddr/corereaders/tdump
+endif
+
+# Exclude source generated by legacy tool, if present.
```

Comment:
There is no legacy tool for Java 11, is this needed?

Reply:
Here, 'legacy' refers to the `PointerGenerator` and `StructureStubGenerator` tools that were in use prior to this change.

Reply:
I'm still not clear on where these legacy classes come from. Are they in git?

Reply:
Prior to this change, `DDR-gensrc.gmk` generated Java source into the source tree under `openj9/debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/structure` and `openj9/debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/pointer/generated`; if any of those files are around, they need to be ignored.

Reply:
So basically this handles re-building a pre-existing build with this change added?

Reply:
Yes.